### ISSUE TITLE
OGG looping music + iOS lock screen bug

### ISSUE THREAD

**Issue by thomasuster** [haxenme/nme#403]
If you play looping OGG looping music on the iPhone, and then use the lock phone hardware button. When you come back in the application it will sometimes cause the application to hang, and other times the application will run but it seems to loop the ogg music silently super fast. Other times it shows the application in a weird resized state and prints this...

```
2016-11-04 10:45:17.389425 AudioTest[3782:2966232] [Common] _BSMachError: port b003; (os/kern) invalid capability (0x14) "Unable to insert COPY_SEND"
2016-11-04 10:45:17.396141 AudioTest[3782:2966232] [Common] _BSMachError: port b003; (os/kern) invalid capability (0x14) "Unable to insert COPY_SEND"
```

**Comment by thomasuster** [comment]
I've discovered this is a regression.

This commit causes the bug.
```
git c 534cdd692749d6557a08e215da071249a076b3b6 -- project/src/audio/OpenALSound.cpp
sudo nme remake ios
```

The previous commit works fine.
```
git c 282535dc23cbe5336c76b66180ec1ea6e627c72f -- project/src/audio/OpenALSound.cpp
sudo nme remake ios
```

**Comment by thomasuster** [comment]
@jskuse Hey there. What did https://github.com/haxenme/nme/commit/534cdd692749d6557a08e215da071249a076b3b6 fix?

**Comment by thomasuster** [comment]
Ahh, I think I get it, I seem to hear a re-trigger of sound effects. But the looping music does seem to work fine.

**Comment by jskuse** [comment]
@thomasuster When an alarm or even (Control Center) interrupted the audio in iOS it resulted in the audio resuming in a mass of noise (basically any active sound would restart from the beginning and play at once, which can be a *lot* because there's a bug whereby "completed" audio channels stay in the active list until GC without the patch in #353). 

We use OGG files too and this has been live in our game for a few months now (and been through a few external QA cycles as well). I've also just tested as per your reproduction steps on our game on the iPad I have to hand (albeit it's an iOS8 device) and it's fine. However we're not running cutting edge NME and we also have #353 applied.

It's the end of the day here so I don't really have the time to test your sample, but can do tomorrow if it'd help.

**Comment by thomasuster** [comment]
@jskuse I'm looking at it today for a bit. So hopefully I can find the right fix. I also tried with https://github.com/haxenme/nme/pull/353. Same issue, I did repo the issue you we working on the the already finished sounds getting re-triggered.

**Comment by jskuse** [comment]
@thomasuster Right you are, well any help I can offer feel free to ask - I know you have a release pending! 

One thing though, are you saying that you repro-ed the issue with this patch (and optionally with 353) in place? Because if so, that *really* doesn't make sense. If so I don't suppose there's a chance that the toolkit build is using a newer version of OpenAL or something?

**Comment by thomasuster** [comment]
Okay, the 353 patch seems to fix your issue without needing the first patch. And without the first patch the bug does not occur. Which makes sense, seems like 353 and other patch seem to be trying to fix the same thing right?

Would you try this?
```
git revert 534cdd692749d6557a08e215da071249a076b3b6
```

For me, since I don't have 353 I tried this.
```
git revert 534cdd692749d6557a08e215da071249a076b3b6
git cherry-pick da89170d02200fa92405e8f662ff5bb1edab7fa2
```

**Comment by thomasuster** [comment]
Not using toolkit for iOS.

**Comment by jskuse** [comment]
@thomasuster Not *quite*. Patch 353 fixed long finished sounds lingering around for ages. This patch fixed current sounds retriggering when the audio resumed after a suspension.

Obviously, with 353 there are fewer current sounds at any one time, which reduces the issue, but this one prevents any sounds that are *currently* playing starting again (we use a lot of narration, so this is very evident in our game). Well, unless something else has changed since I made this change that has fixed it some other way.

**Comment by thomasuster** [comment]
But when they 'replay' wont they replay at their current position? So it should be fine, unless they're not resuming at their current position.

**Comment by jskuse** [comment]
@thomasuster Exactly, they were restarting from the start.

**Comment by jskuse** [comment]
@thomasuster Can I suggest you revert this on your branch for your release and then once that's out of the way we'll put our heads together on gitter and try and work out why it's crashing for your game and not ours?

**Comment by jskuse** [comment]
@thomasuster (...and the more I think about the order I did the fixes, the more I think maybe you're right, it might be that 353 *did* fix them both, because I noticed the channel issue while doing the initial fix in a panic for a release but didn't have time to do it "properly" until later, but I'm out of the office for the night and can't easily check!)

**Comment by thomasuster** [comment]
@jskuse Of course, please prioritize this so you can verify that I'm right or not.

**Comment by thomasuster** [comment]
Hmmm, looks like just those two commits have issues as well.

**Comment by jskuse** [comment]
@thomasuster You mean 353 has issues as well? If so, I can only apologise, although I must reiterate that we've had both of these changes tested throughly and live on the app store with heavy usage for months now without any obvious issues in this area. That *isn't* to say that there aren't issues of course! Just that we've not seen them, maybe just because we're lucky or maybe some crucial difference with our branch / versions of the various dependencies etc. 

If you'd prefer, feel free to revert both changes and I'll just keep them on my branch (since it looks like we're not going to sync any time soon). 

I'm going to be doing some work on the iOS version tomorrow, so if theres anything specific you'd like me to look at (beyond what we discussed re. checking to see if 353 do the job well enough alone) then I'm happy to do so.

**Comment by thomasuster** [comment]
@jskuse Cool, okay.

**Comment by thomasuster** [comment]
This should be fixed with https://github.com/haxenme/nme/pull/406
